modules/dnn/src/caffe/caffe_importer_v2.cpp

```diff
@@ -0,0 +1,579 @@
+            addedBlobs[topName] = LayerPin(layerId, i);
+        }
+    }
+
+    static void extractWeights(const std::string& name,
+                               const ProtobufNode& layersBin, LayerParams& lp)
+    {
+        CV_TRACE_FUNCTION();
+
+        // Find binary layer by name;
+        ProtobufNode layer;
+        for (int i = 0, n = layersBin.size(); i < n; ++i)
+        {
+            if (layersBin[i].has("name") &&
+                name == (std::string)layersBin[i]["name"])
+            {
+                layer = layersBin[i];
+                break;
+            }
+        }
+        if (layer.empty())
+        {
+            return;
+        }
+
+        // Parse blobs.
+        ProtobufNode blobs = layer["blobs"];
+        for (int i = 0, n = blobs.size(); i < n; ++i)
+        {
+            ProtobufNode blob = blobs[i];
+            std::vector<int> blobShape;
+
+            if (blob.has("num") || blob.has("channels") || blob.has("height") ||
+                blob.has("width"))
+            {
+                CV_Assert(blob.has("num") && blob.has("channels") &&
+                          blob.has("height") && blob.has("width"));
+                blobShape.resize(4);
+                blob["num"] >> blobShape[0];
+                blob["channels"] >> blobShape[1];
+                blob["height"] >> blobShape[2];
+                blob["width"] >> blobShape[3];
+            }
+            else
+            {
+                if (blob.has("shape"))
+                {
+                    ProtobufNode shape = blob["shape"]["dim"];
+                    blobShape.resize(shape.size());
+                    for (int i = 0; i < blobShape.size(); ++i)
+                    {
+                        blobShape[i] = (int64_t)shape[i];
+                    }
+                }
+                else  // Is a scalar
+                    blobShape.resize(1, 1);
+            }
+
+            cv::Mat weights(blobShape, CV_32F);
+
+            ProtobufNode blobData = blob["data"];
+            CV_Assert(blobData.size() == weights.total());
+            blobData.copyTo(weights.total() * sizeof(float), weights.data);
+            lp.blobs.push_back(weights);
+        }
+    }
+
+    template <typename T>
+    static void setParam(const std::string& name, const ProtobufNode& node,
+                         LayerParams& lp)
+    {
+        lp.set(name, (T)node[name]);
+    }
+
+    template <typename T>
+    static void setParamArray(const std::string& name,
+                              const ProtobufNode& node, LayerParams& lp)
+    {
+        if (node.has(name))
+        {
+            std::vector<T> param(node[name].size());
+            for (int i = 0; i < param.size(); ++i)
+            {
+                node[name][i] >> param[i];
+            }
+            DictValue dict;
+            if (typeid(T) == typeid(int32_t) || typeid(T) == typeid(uint32_t) ||
+                typeid(T) == typeid(int64_t) || typeid(T) == typeid(uint64_t))
```

Comment:
Use template specializations / type traits instead.
I believe this "if" must be handled by `DictValue` itself.

Reply:
Thanks! Replaced to `is_integer` and `is_specialized` for non-integers but floats and doubles.